[1단계 - 점심 뭐 먹지] 제나(이정은) 미션 제출합니다.

JeLee-river opened the pull request:
안녕하세요, 크리스! 이번에 새롭게 리뷰이가 된 제나입니다. ㅎㅎ 이렇게 리뷰 요청으로 만나뵙게 되어, 정말 반갑습니다! 이번 미션 동안 잘 부탁드립니다! 😄 

---

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://jelee-river.github.io/javascript-lunch

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

이번 미션에서 가장 어려웠던 점은 컴포넌트의 분리였습니다. 저희 페어는 반복되는 로직을 단위로 분리하였습니다. 그러나 `required`와 같이 ui별로 다르게 부여해야 하는 속성들을 추가로 관리해야 했고, 저희가 컴포넌트 분리의 기준을 적절히 설정한 것인지 의문이 들었습니다.

그러나 더 분리한다면, 재사용성에서 의미가 없다고 생각했기 때문에 현재의 코드를 유지하기로 했습니다.

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

<br>

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

### 2-1. 파일의 구조

파일의 구조를 아래와 같이 구성해보았는데, 잘 분리가 된 것일지 피드백을 부탁드리고 싶습니다.

```
📦 src                        # 프로젝트 소스 코드 폴더
 ┣ 📂 components              # 재사용 가능한 UI 컴포넌트 모음
 ┃ ┣ 📜 Button.js            # <button> 요소를 생성하는 컴포넌트
 ┃ ┣ 📜 Header.js            # <header> 요소를 생성하는 컴포넌트 (앱 제목 표시)
 ┃ ┣ 📜 index.js             # components 폴더 내 모든 컴포넌트를 한 곳에서 export
 ┃ ┣ 📜 Input.js             # <input> 요소를 생성하는 컴포넌트 (텍스트 입력 필드)
 ┃ ┣ 📜 Modal.js             # <div class="modal"> 요소를 생성하는 컴포넌트 (모달 UI)
 ┃ ┣ 📜 RestaurantItem.js    # <li> 요소를 생성하는 컴포넌트 (음식점 리스트 항목)
 ┃ ┣ 📜 Select.js            # <select> 요소를 생성하는 컴포넌트 (드롭다운 메뉴)
 ┃ ┗ 📜 Textarea.js          # <textarea> 요소를 생성하는 컴포넌트 (설명 입력 필드)
 ┣ 📜 constants.js           # 변하지 않는 상수 데이터 저장 (카테고리, 거리, 버튼 텍스트 등)
 ┣ 📜 eventHandlers.js       # 사용자 입력 이벤트 핸들러 관리 (모달 열기, 음식점 추가 등)
 ┣ 📜 main.js                # 애플리케이션 진입점 (UI 생성 및 이벤트 핸들러 등록)
 ┣ 📜 restaurantData.js      # 초기 음식점 데이터를 저장하는 파일
 ┣ 📜 stateStore.js          # 애플리케이션 상태 관리 (새 음식점 데이터 저장 및 관리)
 ┣ 📜 style.css              # 애플리케이션의 스타일 정의 (UI 디자인 및 레이아웃)
 📜 index.html             # 웹 애플리케이션의 기본 HTML 구조
```

<br>

### 2-2. required 적용 방식에 대한 고민

현재 아래의 스니펫과 같이 상수 파일에 required 정보를 저장한 후, 이를 컴포넌트에 전달하여 required를 지정해주고 있습니다. 

이와 같은 방식을 선택한 이유는, input 컴포넌트에 required 속성을 부여하기에는 범용성이 떨어진다고 생각했기 때문입니다. 따라서 필수인 input에만  required를 부여하는 방식을 채택했는데, required가 불필요한 input까지 `required: ''`을 전달해주고 있어, 다소 어색하게 느껴지기도 합니다. 이러한 방식에 대한 크리스의 의견이 궁금합니다!


```javascript
// constsnts.js

export const NAME = {
  label: '이름',
  name: 'name',
  helpText: '',
  required: 'form-item--required',
  type: 'text',
};
export const LINK = {
  label: '참고 링크',
  name: 'link',
  helpText: '매장 정보를 확인할 수 있는 링크를 입력해 주세요.',
  required: '',
  type: 'url',
};
```

<br>

### 2-3. HTML 분리를 위한 DOM element 삽입 방식

현재 저희가 작성한 컴포넌트들은 템플릿 리터럴을 사용하여 HTML 문자열을 반환하는 방식으로 구현되어 있습니다.

이 방식의 장점은 각 UI의 구조를 직관적으로 확인할 수 있고, html 분리를 통해 유지보수하기 쉬운 점이라고 생각합니다. 그러나 JS 메서드를 사용하여 직접 DOM element에 템플릿을 삽입해야 하기 때문에, 다른 관점에서는 불필요한 동작을 수행하고 있는 것처럼 느껴지기도 합니다.

이에 대한 크리스의 생각은 어떠신지 궁금합니다. 


<br>

### 2-4. 컴포넌트의 분리

저희 페어는 '반복되는 로직'을 기준으로 컴포넌트를 나누었고, 먼저 Templates 파일을 기반으로 공통된 패턴을 정리한 후 컴포넌트로 분리했습니다.
이 방식이 재사용성과 유지보수 측면에서 적절한 방식이 될 수 있는지 크리스의 의견이 궁금합니다. 또한 크리스의 컴포넌트 기준이 궁금합니다.

<br>

### 2-5. e2e 테스트에서 `{ force: true }` 사용
e2e 테스트 코드를 작성하던 중, cypress가 modal을 인식하지 못하는 문제로 인해, 다음과 같이 코드에 `{ force: true }` 속성을 부여했습니다.

이렇게 요소가 안 보이더라도 강제 클릭하게 하는 기능을 사용하여, 테스트를 통과하는 것이 적절한 방식일지 고민입니다.
크리스의 의견은 어떠신가요?

```javascript
 it('사용자가 음식점 추가 모달을 연 후, ESC 키를 클릭하여 음식점 추가를 취소하는 시나리오 테스트', () => {
    cy.get('.gnb__button').click();
    cy.get('.modal--open').should('exist');

    cy.document().trigger('keydown', { key: 'Escape', keyCode: 27, which: 27, force: true });
    cy.get('.modal--open').should('not.exist');
  });
```

<br>


---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

JeLee-river committed:
chore: eslint, prettier 설정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 기능 목록 정리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
chore: 이미지 파일을 public으로 이동

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 전체 레이아웃 및 스타일 적용

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 동적으로 GNB 헤더 생성 및 페이지 로드시 추가

- createHeader 함수 추가하여 헤더 동적 생성
- window load 이벤트를 사용해 페이지 로드시 헤더 자동 삽입
- 기존 정적인 HTML 대신 JavaScript로 GNB 구조 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 전체 레이아웃 중앙 정렬

- body와 html에 flexbox를 적용하여 전체 페이지 중앙 정렬

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 헤더 생성 구현 완료에 따른 구현할 기능 목록 업데이트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 동적으로 음식점 아이템 리스트 생성 및 페이지 로드 시 추가

- 페이지 로드 시 음식점 리스트를 동적으로 생성하여 `.restaurant-list`에 추가
- `RESTAURANTS` 데이터를 기반으로 `createRestaurantItem` 함수를 활용하여 리스트 아이템을 생성
- `insertAdjacentHTML('beforeend', items)`를 사용하여 기존 리스트에 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 테스트 목록 작성

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 상수를 constants.js 파일로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: app의 주요 ui 요소 렌더링 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 음식점 추가 모달을 여는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 새 음식점 정보를 받아와서 저장하는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달의 추가하기 버튼을 누르면 음식점 아이템을 추가하는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 취소하기 버튼을 누르면 모달을 닫는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: main.js에서 lunchUI.js로 ui 로직 분리

- main.js는 애플리케이션 로직을 담당하도록 유지
- UI 관련 함수는 lunchUI.js에서 관리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 취소하기 버튼, 추가하기 버튼을 눌렀을 때 입력을 초기화 하고 닫는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
style: 모달 너비 조정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달 외부를 클릭했을때 입력을 초기화하고 닫는 기능 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
fix: enter, space 키를 눌러도 모달이 닫히는 현상 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 구현 완료한 기능 체크

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: 모달창 UI 구현

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 모달 렌더링 테스트 추가

JeLee-river committed:
test: constants.js 파일 변경에 따른 테스트 로직 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: e2e 테스트 목록 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 열어, 모든 필드를 입력 후 음식점을 추가하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 열어, 필수 입력값만 입력 후 추가하기 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 음식점 추가 중 취소하기 버튼을 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

JeLee-river committed:
test: modal이 열리는지 확인하는 로직 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, ESC 키를 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 모달 외부를 클릭하여 음식점 추가를 취소하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 음식점 추가 모달을 연 후, 모달을 닫았다가 다시 열었을 때 필드가 초기화 되는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
docs: 모달이 렌더링 되는지 확인하는 렌더링 테스트 체크

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
feat: input link의 type을 url로 설정

- 그에 따른 테스트 코드 변경
- constants.js에 type 속성 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
test: 모달의 필수 필드 입력창에 아무것도 입력하지 않고 추가하기 버튼을 클릭하는 시나리오 테스트 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: lunchUI를 컴포넌트별로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 상수 파일에서 레스토랑 데이터를 restaurantData.js로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 거리 데이터를 숫자형으로 변경

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: UI 생성 로직을 함수로 분리하여 코드 가독성 개선

- 헤더 추가 로직을 appendHeader()로 분리
- 음식점 리스트 초기화 로직을 initRestaurantItems()로 분리
- 모달 생성 로직을 appendModal()로 분리
- 모달 내부 컨텐츠 추가 로직을 appendModalContents()로 분리
- 버튼 추가 로직을 appendModalButton()으로 분리
- 이벤트 핸들러 등록 로직을 addEventHandlers()로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
refactor: 버튼 추가 로직을 추가로 분리

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

JeLee-river committed:
chore: 배포 설정

JeLee-river committed:
fix: 사진이 배포 안되는 문제 수정

swon3210 commented on src/eventHandlers.js:
```diff
@@ -0,0 +1,80 @@
+  });
+}
+
+function resetForm() {
+  const form = document.querySelector('#new-restaurant-form');
+  form.reset();
+}
```
input 컴포넌트 하나하나에 직접 접근하는 방식이 아니라 form 을 이용해서 한번에 초기화하는 방식을 쓰셨군요. 불필요한 코드를 줄이는 매우 효율적인 방식이라고 생각됩니다!

swon3210 reviewed (approved):
전체적으로 매우 잘 구성된 코드입니다. 컴포넌트간의 의존성이 매우 적으며 주요 로직은 최상위 컴포넌트(main)에서 관리되고 있습니다. 컴포넌트의 재사용성도 매우 높다고 볼 수 있습니다

swon3210 commented:
### >2-1. 파일의 구조

>파일의 구조를 아래와 같이 구성해보았는데, 잘 분리가 된 것일지 피드백을 부탁드리고 싶습니다.

```
📦 src                        # 프로젝트 소스 코드 폴더
 ┣ 📂 components              # 재사용 가능한 UI 컴포넌트 모음
 ┃ ┣ 📜 Button.js            # <button> 요소를 생성하는 컴포넌트
 ┃ ┣ 📜 Header.js            # <header> 요소를 생성하는 컴포넌트 (앱 제목 표시)
 ┃ ┣ 📜 index.js             # components 폴더 내 모든 컴포넌트를 한 곳에서 export
 ┃ ┣ 📜 Input.js             # <input> 요소를 생성하는 컴포넌트 (텍스트 입력 필드)
 ┃ ┣ 📜 Modal.js             # <div class="modal"> 요소를 생성하는 컴포넌트 (모달 UI)
 ┃ ┣ 📜 RestaurantItem.js    # <li> 요소를 생성하는 컴포넌트 (음식점 리스트 항목)
 ┃ ┣ 📜 Select.js            # <select> 요소를 생성하는 컴포넌트 (드롭다운 메뉴)
 ┃ ┗ 📜 Textarea.js          # <textarea> 요소를 생성하는 컴포넌트 (설명 입력 필드)
 ┣ 📜 constants.js           # 변하지 않는 상수 데이터 저장 (카테고리, 거리, 버튼 텍스트 등)
 ┣ 📜 eventHandlers.js       # 사용자 입력 이벤트 핸들러 관리 (모달 열기, 음식점 추가 등)
 ┣ 📜 main.js                # 애플리케이션 진입점 (UI 생성 및 이벤트 핸들러 등록)
 ┣ 📜 restaurantData.js      # 초기 음식점 데이터를 저장하는 파일
 ┣ 📜 stateStore.js          # 애플리케이션 상태 관리 (새 음식점 데이터 저장 및 관리)
 ┣ 📜 style.css              # 애플리케이션의 스타일 정의 (UI 디자인 및 레이아웃)
 📜 index.html             # 웹 애플리케이션의 기본 HTML 구조
```

적절하게 분리된 구조라고 생각합니다. 컴포넌트들이 서로 의존하지 않으며 주요 로직은 최상위 컴포넌트(main)에서 관리되고 있습니다. 모든 파일이 독립된 책임을 지니고 있으므로 재사용성도 매우 높다고 볼 수 있습니다.

### >2-2. required 적용 방식에 대한 고민

>현재 아래의 스니펫과 같이 상수 파일에 required 정보를 저장한 후, 이를 컴포넌트에 전달하여 required를 지정해주고 있습니다.

>이와 같은 방식을 선택한 이유는, input 컴포넌트에 required 속성을 부여하기에는 범용성이 떨어진다고 생각했기 때문입니다. 따라서 필수인 input에만 required를 부여하는 방식을 채택했는데, required가 불필요한 input까지 `required: ''`을 전달해주고 있어, 다소 어색하게 느껴지기도 합니다. 이러한 방식에 대한 크리스의 의견이 궁금합니다!

```
// constsnts.js

export const NAME = {
  label: '이름',
  name: 'name',
  helpText: '',
  required: 'form-item--required',
  type: 'text',
};
export const LINK = {
  label: '참고 링크',
  name: 'link',
  helpText: '매장 정보를 확인할 수 있는 링크를 입력해 주세요.',
  required: '',
  type: 'url',
};
```

매우 적절한 방식이라고 생각합니다. 해당 객체들은 사실상 Input 컴포넌트에 넘기는 매개변수들의 모음으로 보이는데요. Input 컴포넌트를 사용하는 방식을 사용처마다 유연하게 선택할 수 있는 인터페이스를 제공해주는 것이므로 Input 컴포넌트의 재사용성을 높이는 방식이라고 볼 수 있습니다.

다만 required 에 특정한 형식의 문자열을 지정하는게 필수는 아닌 것 같습니다. required 값은 true 혹은 false 를 받거나, 아예 받지 않도록 설정할 수 있습니다. 예를 들어 다음과 같이 Input 컴포넌트를 수정할 수 있습니다.

```jsx
function createInput(fieldName) {
  const input = `<div class="form-item ${fieldName.required ? 'form-item--required' : ''}">
    <label for="${fieldName.name} text-caption">${fieldName.label}</label>
    <input type="${fieldName.type}" name="${fieldName.name}" id="${fieldName.name}">
    <span class="help-text text-caption">${fieldName.helpText}</span>
  </div>`;

  return input;
}

export default createInput;

```

### >2-3. HTML 분리를 위한 DOM element 삽입 방식

>현재 저희가 작성한 컴포넌트들은 템플릿 리터럴을 사용하여 HTML 문자열을 반환하는 방식으로 구현되어 있습니다.

>이 방식의 장점은 각 UI의 구조를 직관적으로 확인할 수 있고, html 분리를 통해 유지보수하기 쉬운 점이라고 생각합니다. 그러나 JS 메서드를 사용하여 직접 DOM element에 템플릿을 삽입해야 하기 때문에, 다른 관점에서는 불필요한 동작을 수행하고 있는 것처럼 느껴지기도 합니다.

>이에 대한 크리스의 생각은 어떠신지 궁금합니다.

현재 요구사항이 매우 단순하기 때문에, 관점에 따라서는 불필요한 동작처럼 느껴질 수 있습니다. 렌더링의 상당 부분을 HTML/CSS 에 맡기고 지금의 HTML 문자열 렌더링 로직을 제거하면 실제로 코드를 지금보다 훨씬 줄일 수 있을 것입니다.

그러나 요구사항이 여기서 좀 더 복잡해지고, 이에 따라 렌더링 분기가 늘어난다면 지금의 렌더링 방식이 빛을 발할 수 있습니다. 리액트와 마찬가지로, 렌더링 영역을 쉽게 나눌 수 있고, DOM 에 직접 접근하지 않고도 렌더링 분기를 상태 혹은 매개변수에 따라 간편하게 구현할 수 있기 때문입니다.

결국 모든 것의 기준은 서비스 요구사항에 달려 있습니다. 대부분의 경우, 제나의 방식이 더 효용이 높습니다. 다만 저는 미리 앞날을 생각하면서 개발하기 보다는 당장의 요구사항에 집중하는 것이 생산성 측면에서 훨씬 효율적이라고 생각하기에, 저라면 HTML/CSS 에 렌더링의 상당수를 맡기는 것을 선택할 것 같습니다.

물론, 제나가 현재의 코드를 수정하실 필요는 전혀 없습니다.

### 2-4. 컴포넌트의 분리

>저희 페어는 '반복되는 로직'을 기준으로 컴포넌트를 나누었고, 먼저 Templates 파일을 기반으로 공통된 패턴을 정리한 후 컴포넌트로 분리했습니다.
>
>이 방식이 재사용성과 유지보수 측면에서 적절한 방식이 될 수 있는지 크리스의 의견이 궁금합니다. 또한 크리스의 컴포넌트 기준이 궁금합니다.

제나가 진행하신 방식이 유지보수 측면에서 적절한 방식이 맞습니다. ‘컴포넌트 기준’이라 함은 컴포넌트를 어떤 식으로 구성하는지를 여쭤보는 것으로 이해하면 될까요? 그 질문이 맞다면, 저는 전체 요구사항을 ‘책임’이라는 단위로 나눈다고 말씀드릴 수 있습니다.

예를들어, ‘이용자가 입력한 음식점 정보를 리스트에 추가한다’ 라는 요구사항이 있다면, 이 요구사항을 ‘이용자의 입력을 받는다.’ 와 ‘리스트에 음식을 추가한다’ 라는 두가지 책임으로 나눌 수가 있습니다.

이 경우, ‘이용자의 입력을 받는’ RestaurantAddForm 컴포넌트를 만들 수 있고 RestaurantAddForm 컴포넌트에서 받아온 입력값을 바탕으로 리스트에 음식점을 추가하는 ‘LunchRestaurantManger’ 상위 컴포넌트를 만들 수 있을 것입니다.

여기서 컴포넌트가 ‘하나의 책임’ 만을 가지고 있도록 하면, 재사용성을 갖출 수 있습니다. 예를 들어 RestaurantAddForm 컴포넌트가 단순히 사용자 입력을 받는 것 외에도 음식점을 리스트에 추가하는 책임까지 맡게 되면, 해당 컴포넌트를 다른 페이지에서 다른 맥락으로 활용하기 어려워질 것입니다.

예를 들어, 점심 뭐먹지 기능에 더불어 ‘저녁 뭐먹지' 기능이 추가된다면, RestaurantAddForm 가 입력을 받는 역할만을 할 때, 저녁 음식점 리스트에 음식점을 추가하는 데에도 사용될 수 있을 것입니다. 하지만 RestaurantAddForm 컴포넌트가 점심 음식점 추가 책임을 지니고 있다면 이를 재사용할 수 없으니 또 다른 컴포넌트를 새로 만들거나, RestaurantAddForm 의 책임을 하나 더 늘려야 합니다. RestaurantAddForm 이 너무 많은 책임을 가지게 될 시, 재사용성은 더욱 하락하며 유지보수 관점에서도 좋지 않은 코드가 될 것입니다.

### 2-5. e2e 테스트에서 `{ force: true }` 사용

>e2e 테스트 코드를 작성하던 중, cypress가 modal을 인식하지 못하는 문제로 인해, 다음과 같이 코드에 `{ force: true }` 속성을 부여했습니다.
>
>이렇게 요소가 안 보이더라도 강제 클릭하게 하는 기능을 사용하여, 테스트를 통과하는 것이 적절한 방식일지 고민입니다.
>
>크리스의 의견은 어떠신가요?

```
 it('사용자가 음식점 추가 모달을 연 후, ESC 키를 클릭하여 음식점 추가를 취소하는 시나리오 테스트', () => {
    cy.get('.gnb__button').click();
    cy.get('.modal--open').should('exist');

    cy.document().trigger('keydown', { key: 'Escape', keyCode: 27, which: 27, force: true });
    cy.get('.modal--open').should('not.exist');
  });
```

아뇨 좋은 방식이 아니라고 생각합니다. 이용자는 보이지 않는 요소와 상호작용하지 못하기에, 보이지 않더라도 이벤트를 트리거 시키는 force: true 방식은 최대한 지양해야 합니다. 제가 잘못 실행한게 아니라면, 해당 옵션을 제거해도 테스트가 통과하는 것으로 보이는데요. 제가 알기로는 document 는 실제 가시성 여부와 상관없이 무조건 상호작용 가능한 대상으로 인식하기 때문입니다. 혹시 정확히 어떤 문제가 있었는지 설명 부탁드릴 수 있을까요? 아니면 제가 어딘가 놓친 부분이 있을까요?

swon3210 reviewed (changes requested):
전반적인 요구사항을 모두 만족하고 있으나, 테스트 코드에서 force: true 옵션이 과하게 사용되고 있는 것으로 보여 Request changes 로 바꾸도록 하겠습니다. 제가 볼 때는 force: true 옵션을 충분히 없앨 수 있을 것으로 보이는데, 혹시 문제가 있다면 말씀주시면 감사드리겠습니다.

<img width="311" alt="스크린샷 2025-03-08 오후 4 01 39" src="https://github.com/user-attachments/assets/0e6c074c-79ef-4b89-82ee-64ea57fedc03" />

일단 기존의 css 에 문제가 있어 작은 화면에서는 실제로 보이지 않는 요소가 생긴다는 것이 문제로 보입니다. 모달에 스크롤을 넣거나, 레이아웃을 조정하는 방식으로 해당 문제를 수정할 수 있을 것으로 보입니다.

JeLee-river committed:
test: cypress viewport를 조정해 요소의 테스트가 정상적으로 수행되도록 수정

JeLee-river committed:
test: cypress가 backdrop 클릭하지 못하는 문제 수정

- 모달이 z-index 상으로 최상단에 존재하여, backdrop를 클릭해도 모달이 클릭되는 문제로 인해 테스트에 실패했음

- 테스트 도중에 모달 외부 요소의 z-index를 조정하여 cypress가 backdrop를 클릭하도록 수정

JeLee-river commented:
> 전반적인 요구사항을 모두 만족하고 있으나, 테스트 코드에서 force: true 옵션이 과하게 사용되고 있는 것으로 보여 Request changes 로 바꾸도록 하겠습니다. 제가 볼 때는 force: true 옵션을 충분히 없앨 수 있을 것으로 보이는데, 혹시 문제가 있다면 말씀주시면 감사드리겠습니다.
> 

<br>

리뷰 감사드립니다! 남겨주신 코멘트를 토대로 원인을 파악하고 다음과 같이 수정했습니다.


### 1. 뷰포트 설정 문제

- select의 `cy.click()`과 modal의 `keydown` 이벤트에서 발생한 에러는 작은 화면에서 요소가 보이지 않는 점이 원인이었습니다.

- **수정 내용**
  - cypress의 뷰포트 설정을 조정하여, 화면에서 모든 요소가 보이도록 했습니다. 뷰포트 크기는 [cypress의 공식문서](https://docs.cypress.io/api/commands/viewport#Arguments)에 작성된 preset 중 가장 큰 크기로 임시 설정하였습니다.

  - 세부 변경 사항은 다음과 같습니다: [72a02be](https://github.com/woowacourse/javascript-lunch/pull/217/commits/72a02be22a03ba5ffed268acd7f4466d7efd2c20)
  
  ```javascript
  // cypress.config.ts
  
  ...
  
  export default defineConfig({
    e2e: {
      viewportWidth: 1536,
      viewportHeight: 960,
    },
  
  ...
  });
  
  ```

<br>

### 2. z-index로 인한 클릭 문제

- 모달의 `cy.click()`에서 발생한 에러는 backdrop의 z-index가 모달보다 낮아, cypress가 backdrop을 클릭할 때 상단에 위치한 모달이 대신 클릭되는 문제였습니다.

- **수정 내용**

  - `force: true` 옵션을 사용하지 않고 문제를 해결하기 위해 cypress가 z-index 조정을 통해 `.modal-backdrop`에 접근할 수 있도록 수정했습니다.
  
  - 세부 변경 사항은 다음과 같습니다: [de56fa5](https://github.com/woowacourse/javascript-lunch/pull/217/commits/de56fa55f33a01884a3675f04260829645411310)
  
  ```javascript
   cy.get('.modal-backdrop').invoke('css', 'z-index', '9999').click(); // index 조정
   cy.get('.modal--open').should('not.exist');
  ```

JeLee-river committed:
test: 테스트 설명의 통일성을 위해 문자열 수정

JeLee-river committed:
test: 가독성을 고려하여 테스트 종류별로 블록 분리

JeLee-river committed:
refactor: 음식점 아이템 데이터의 required 속성을 boolean으로 수정

JeLee-river commented:
## 추가 질문 및 코멘트

> ### 2-2. required 적용 방식에 대한 고민
> >
> 다만 required 에 특정한 형식의 문자열을 지정하는게 필수는 아닌 것 같습니다. required 값은 true 혹은 false 를 받거나, 아예 받지 않도록 설정할 수 있습니다.

사실, required 속성은 boolean 타입으로 설정하고, 이에 따라 삼항 연산자로 className을 추가하는 방식을 고려했었습니다. 그러나 우테코의 기존 요구사항인 '삼항 연산자를 사용하지 않는다'는 조건이 마음에 걸려 고민 끝에 문자열 지정 방식을 선택하게 되었습니다.

일부 상황에서는 삼항 연산자가 코드의 가독성을 떨어뜨릴 수 있기 때문에 이러한 요구사항이 설정된 것이라고 생각합니다. 하지만 현재 상황과 같이 구현한 코드와 요구사항의 의도가 잘 맞지 않는다고 느낄 때, 요구사항을 과감히 무시하고 코드 스타일을 바꿀지, 아니면 요구사항을 지키기 위해 억지로라도 따를지에 대해 미션을 진행하면서 계속 고민이 됩니다. 😂

우테코 수료생으로서 크리스의 의견이 궁금합니다! ㅎㅎ

<br>
<br>

> ### >2-3. HTML 분리를 위한 DOM element 삽입 방식
> >

프로젝트의 규모와 확장성, 요구사항에 따라 달라질 수 있군요..! 답변 감사합니다. 👍 

그렇게면 혹시, 현재의 코드처럼 1. templates를 만들어 내장 메서드로 추가하는 방식과, 2. `createElement`로 직접 DOM을 생성해서 추가하는 방식 중 어느 방식을 더 선호하시는지 궁금합니다.

이번 미션에서는 엘리먼트의 구조를 한 번에 알 수 있는 직관성을 고려하여 templates를 사용했으나, 안전성 측면에서는 `createElement`으로 직접 DOM을 만들어 추가하는 방식이 더 좋다는 점을 알게 되었습니다. 또한 `createElement` 방식으로도 충분히 직관적인 코드를 작성할 수 있다는 의견을 들었습니다. 

제가 고민해 본 결과, `createElement`로 DOM을 생성하여 상위 엘리먼트에 추가하는 일련의 로직을 여러 함수로 분리한다면 가독성과 안전을 모두 잡을 수 있지 않을까 하는 생각이 들었습니다. 이에 대한 크리스의 의견도 궁금합니다. 혹시 더 좋은 방법이 있을까요?

<br>
<br>

> ### 2-5. e2e 테스트에서 `{ force: true }` 사용
> >
> 아뇨 좋은 방식이 아니라고 생각합니다. 이용자는 보이지 않는 요소와 상호작용하지 못하기에, 보이지 않더라도 이벤트를 트리거 시키는 force: true 방식은 최대한 지양해야 합니다. 제가 잘못 실행한게 아니라면, 해당 옵션을 제거해도 테스트가 통과하는 것으로 보이는데요. 제가 알기로는 document 는 실제 가시성 여부와 상관없이 무조건 상호작용 가능한 대상으로 인식하기 때문입니다.

크리스의 코멘트로 힌트를 얻어, 코드를 수정했습니다! 감사합니다! ㅎㅎ

https://github.com/woowacourse/javascript-lunch/pull/217#issuecomment-2708623270

<br>

## 변경 사항

피드백을 반영하여 수정한 내용입니다.

### 1. 기능 테스트의 `{ force: true }` 옵션 제거

 - cypress의 viewport 크기를 조정 (https://github.com/woowacourse/javascript-lunch/pull/217/commits/72a02be22a03ba5ffed268acd7f4466d7efd2c20)
 - 테스트 상에서 모달 backdrop의 z-index를 변경 (https://github.com/woowacourse/javascript-lunch/pull/217/commits/de56fa55f33a01884a3675f04260829645411310)

<br>

### 2. 테스트 설명 수정

- 테스트 설명의 통일성을 위해 문구 수정 (https://github.com/woowacourse/javascript-lunch/pull/217/commits/dac54e0b1b44108d3eef36d190a4a216bb32a089)

<br>

### 3. 유사한 테스트를 블록별로 분리

- 가독성을 고려하여 테스트의 종류별로 블록을 분리 (https://github.com/woowacourse/javascript-lunch/pull/217/commits/ad05cbcb2439045093f189990f6cacfd6cf10fa3)

<br>

### 4. 음식점 아이템 데이터의 `required` 속성을 `boolean` 타입으로 수정

- `required` 여부를 문자열로 전달하는 것보다 boolean 값을 넘겨주어 템플릿에서 적절하게 분기 처리 하는 것이 적절하다 판단
- `required` 속성값을 boolean으로 수정하고 템플릿의 required class를 삼항연산자로 분기 처리 (https://github.com/woowacourse/javascript-lunch/pull/217/commits/b87a71558e48cb9dba1f4392754fa09cdc77cb59)

swon3210 commented:
[ 추가 질문에 대한 답변 ]

>사실, required 속성은 boolean 타입으로 설정하고, 이에 따라 삼항 연산자로 className을 추가하는 방식을 고려했었습니다. 그러나 우테코의 기존 요구사항인 '삼항 연산자를 사용하지 않는다'는 조건이 마음에 걸려 고민 끝에 문자열 지정 방식을 선택하게 되었습니다.

>일부 상황에서는 삼항 연산자가 코드의 가독성을 떨어뜨릴 수 있기 때문에 이러한 요구사항이 설정된 것이라고 생각합니다. 하지만 현재 상황과 같이 구현한 코드와 요구사항의 의도가 잘 맞지 않는다고 느낄 때, 요구사항을 과감히 무시하고 코드 스타일을 바꿀지, 아니면 요구사항을 지키기 위해 억지로라도 따를지에 대해 미션을 진행하면서 계속 고민이 됩니다. 😂

>우테코 수료생으로서 크리스의 의견이 궁금합니다! ㅎㅎ

저도 크루 시절엔 우테코에서 강조한 대로 삼항연산자를 최대한 쓰지 않는 것에 집착했던게 기억이 납니다. 하지만 현업에 와서 삼항 연산자가 정말 많이 쓰이는 것을 알게 됬고, 그 효용도 체감했습니다. 

삼항연산자는 분기를 inline 으로 집어넣을 수 있다는 점에서 매우 큰 강점을 가지고 있습니다. 제가 예시로 든 템플릿 문자열 내에서, 혹은 앞으로 사용하시게 될 리액트 컴포넌트 내에서 분기를 넣을 때, 삼항연산자는 별도의 함수를 정의하지 않아도 분기를 넣을 수 있게 해줍니다. 

물론 논리가 조금만 복잡해져도 가독성이 크게 떨어지고 이로 인해 실수가 발생할 가능성도 있습니다.  우테코에서는 이를 우려하는 것이고, 기준을 잡기가 애매하기 때문에 사용 자체를 지양하도록 권고하는 것이라 생각합니다. 하지만 제 생각에, 삼항연산자는 아주 단순한 비교 연산에서는 가독성을 크게 떨어뜨리지 않고, 오히려 코드를 이해하기 쉽게 해주는 경우도 많습니다.

현업에서도 요구사항이 그것의 실제 의도와 맞지 않는 경우가 꽤 있습니다. 그때 개발자의 역할은 본인이 생각하는 실제 의도에 대해 설명하고, 이에 맞는 다른 방안을 제안하는 것이라 생각합니다. 따라서 요구사항을 ‘과감히 무시’하기 보다는, 직접 제나가 생각하는 삼항연산자를 제한하는 의도에 대해 우테코 측에 설명해보고, 특정한 경우엔 사용하는 것을 허용하도록 제안하는 것을 권해보고 싶습니다. 현장에서 필요로 하는 능력을 함양할 기회가  될 수 있지 않을까요? 물론, 개인적인 제안일 뿐이며 부담이 되신다면 그렇게 하지 않아도 괜찮습니다.


>그렇게면 혹시, 현재의 코드처럼 1. templates를 만들어 내장 메서드로 추가하는 방식과, 2. `createElement`로 직접 DOM을 생성해서 추가하는 방식 중 어느 방식을 더 선호하시는지 궁금합니다.

>이번 미션에서는 엘리먼트의 구조를 한 번에 알 수 있는 직관성을 고려하여 templates를 사용했으나, 안전성 측면에서는 `createElement`으로 직접 DOM을 만들어 추가하는 방식이 더 좋다는 점을 알게 되었습니다. 또한 `createElement` 방식으로도 충분히 직관적인 코드를 작성할 수 있다는 의견을 들었습니다.

>제가 고민해 본 결과, `createElement`로 DOM을 생성하여 상위 엘리먼트에 추가하는 일련의 로직을 여러 함수로 분리한다면 가독성과 안전을 모두 잡을 수 있지 않을까 하는 생각이 들었습니다. 이에 대한 크리스의 의견도 궁금합니다. 혹시 더 좋은 방법이 있을까요?

`createElement` 을 사용하는 함수를 만드는 것이 저 또한 좋은 방법이라 생각됩니다. 이때, HTML 문자열과 동일하게 엘리먼트들의 구조를 나타낼 수 있는 함수를 만들어야 한다고 생각합니다. 이를 위해 해당 함수가 만들어내는 Element 의 children 도 같이 명시할 수 있도록 하는 것을 생각해볼 수 있습니다.

```jsx
function appendModalContents() {
  const form = document.querySelector('#new-restaurant-form');
  const categorySelect = createSelect(CATEGORY);
  const nameInput = createInput(NAME);
  const distanceSelect = createSelect(DISTANCE);
  const descriptionTextarea = createTextarea(DESCRIPTION);
  const linkInput = createInput(LINK);

  form.insertAdjacentHTML('beforeend', categorySelect);
  form.insertAdjacentHTML('beforeend', nameInput);
  form.insertAdjacentHTML('beforeend', distanceSelect);
  form.insertAdjacentHTML('beforeend', descriptionTextarea);
  form.insertAdjacentHTML('beforeend', linkInput);

  appendModalButton(form);
}

function appendModalButton(form) {
  const buttonDiv = document.createElement('div');
  buttonDiv.classList.add('button-container');
  form.appendChild(buttonDiv);

  const addButton = createButton(ADD_BUTTON);
  const cancelButton = createButton(CANCEL_BUTTON);

  const buttonContainer = document.querySelector('.button-container');
  buttonContainer.insertAdjacentHTML('beforeend', cancelButton);
  buttonContainer.insertAdjacentHTML('beforeend', addButton);
}
```

위의 코드를 예시로 들어본다면, 아래처럼 변형하는 것을 의미합니다. (단순히 이해를 돕기 위한 코드이니 추가적으로 리팩터링할 부분이 보인다면 해주시면 됩니다)

```jsx
function createElement(tagName, attributes, children) {
  const element = document.createElement(tagName);
  Object.entries(attributes).forEach(([key, value]) => {
    element[key] = value;
  });

  if (Array.isArray(children)) {
    children.forEach((child) => {
      element.appendChild(child);
    });
  } else {
    element.appendChild(children);
  }

  return element;
}

function createModalForm() {
  return createElement('form', { id: 'new-restaurant-form' }, [
    createSelect(CATEGORY),
    createInput(NAME),
    createSelect(DISTANCE),
    createTextarea(DESCRIPTION),
    createInput(LINK),
    createModalButtons(),
  ]);
}

function createModalButtons() {
  return createElement('div', { class: 'button-container' }, [
    createButton(ADD_BUTTON),
    createButton(CANCEL_BUTTON),
  ]);
}
```

swon3210 reviewed (approved):
테스트 코드에서의 force: true 옵션이 모두 제거된 것을 확인했고, 말씀하신 개선사항들도 모두 충분히 유효한 개선이라고 생각됩니다.

개선사항들을 커밋별로 나눠주셔서 확인하는 것이 훨씬 편리했습니다! 감사합니다 제나!

전반적인 요구사항들을 모두 만족하고 있는 것으로 보이므로 어프로브하겠습니다. 추가적인 질문이 없다면 추후 머지하도록 하겠습니다.
